### PR TITLE
Add Python Style Guide

### PR DESCRIPTION
**Pull request by richardjonathonharris**
This is the alpha draft of a Python style guide for Ascent. I'd also like to put out a second guide for Python as applied to data science libraries (primarily Numpy, Pandas, scikit-learn, and Keras) in the same vein as the rails guide works with the ruby guide here.

Most of the material here is the same as found in PIP 8, but reorganized slightly.

### CODE REVIEW

File: python.md [hunk 1]
```diff
@@ -0,0 +1,309 @@
+# Python
+
+Much of this style guide directly follows [PEP 8](http://pep8.org/). For cases that are not covered here, we encourage you to follow those guidelines. Additionally, [PEP 20](https://www.python.org/dev/peps/pep-0020/) contains some helpful pointers as well.
+
+## General Coding Style
+
+* 4 spaces per indentation level
+    * PyCharm: Preferences -> Code Style -> Python -> Tabs and Spaces
+    * Vim: `set tabstop=4` in `.vimrc`
+* (Roughly) 80 characters per line
```
Comment (danilito19) [inline python.md]: sooo.... I very frequently break this rule. In fact, my rubymine is set to 120 and I still break that. I feel that if we're going to try to follow this, then we need to give some advice as to how to actually stay within it. Is the problem that variable / function names are too long? Is the problem that there's too much indentation already (nested loops, if statements), or that breaking into a second line just looks plain bad to me?

File: python.md [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    * Depending on the context, this can be broken
+    * In general, insert line breaks in between parentheses, brackets, and braces to divide longer statements
+    * However, using a backslash to divide lines is acceptable where there is no other option (such as within a context manager or `assert` statement).
+    
+```python
+# In general, insert line breaks to divide functions
+def some_very_long_function_with_lots_of_arguments(argument1,
```
Comment (danilito19) [inline python.md]: this is an example where I would surpass 80 chars, bc those arguments are very short worded so it looks weird to have them on separate lines. thoughts?

File: python.md [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+# But using a \ is also allowed in some cases such as context managers
+
+with open('this/directory/is/very/long/youknow.txt', 'rb') as fileone, \
+    open('this/other/directory/is/also/really/long/youknow.txt', 'rb') as filetwo:
+    
+    pass
+```
+
+* 2 lines between top level functions and classes, 1 line between methods within a class
+* While single and double quotes to denote strings are both acceptable, keep their use consistent across the project
```
Comment (danilito19) [inline python.md]: I agree with this. However, I have seen within the same project that certain strings seem to use single quotes, like `'x'` or anything short like that. Just throwing that out there. I usually prefer double.

File: python.md [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+def under_10(value):
+    if value < 10:
+        return True
+    else:
+        return None
+        
+# not good
+def under_10(value):
+    if value < 10:
+        return True
```
Comment (danilito19) [inline python.md]: this is actually my preferred way of doing it. I think that returning None is redundant given that most python developers know that a method will always return None.

Comment (danilito19) [reply python.md]: I think the bigger problem is the fact that one return returns `True` while the other returns `None`, which are different types. I'd almost prefer returning False to match the types. We should always try to return the same type of value, and in general, avoid returning None. So maybe here is a good section to say something like "despite above example, avoid returning None if possible"

File: python.md [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+```
+* To check if a collection is empty, make use of the fact that an empty collection evaluates to `False`:
+```python
+# good
+if my_list:
+
+# not good
+if len(my_list) > 0:
+```
+* Avoid comparing `True` or `False` evaluations using `==`:
```
Comment (danilito19) [inline python.md]: I'd move this section up to where you talk about comparing `is None`

File: python.md (hunk 1 above)
Comment (chelseatroy) [reply python.md]: I also break this rule pretty frequently. Historically, this rule is a skeuomorph from the 1928 IBM punch card, which only had 80 columns. Given that the technology is more advanced such that that is no longer the case, I'd argue that this rule has much less value now than when it was first introduced. 

There is still value to having a line length recommendation: human eyes read best when the line is about 5 inches wide (narrower for speed reading, but we don't speed read code). Shorter lines also make it possible to examine two files side by side. 

In general, I find 100-120 characters to be comfortable for fitting the above strictures. Below that, developers and up inventing weird ways to break up the code to get under the line recommendation. Above it, and I cannot put files side by side and hope to read both sides without needing to scroll right.

I'm not particularly invested in whether we go with 80 here or not, but I thought I'd provide the background on the original argument for 80 as well as the arguments now :)

File: python.md [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+* However, if required, explicit relative imports are allowed (`from .models import ReferenceDocument`)
+* Specific libraries sometimes have specific ways to import them:
+    * Import `pandas` and `numpy` as `pd` and `np` respectively (`import pandas as pd`)
+    * Import individual objects from their modules in `scikit-learn` and `keras` (`from sklearn.linear_models import LogisticRegression`, `from keras.layers import Dense`)
+    * When applicable but not detailed here, try to import third-party libraries in the same style that their docs do (such as `nlp = spacy.load('en')` or `import matplotlib.pyplot as plt`)
+
+## Whitespace
+
+* **Add one space before and after assignment (`=`) or any other operator (`a = b`)**
+    * However, no spaces surrounding `=` for a keyword argument (`a(param=b)` not `a(param = b)`)
```
Comment (chelseatroy) [inline python.md]: see review

File: python.md [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        * This will prevent the method from being imported through `from MyClass import *`
+    * Starting the name with two underscores `__` will invoke name mangling to avoid clashes with any subclasses:
+        * For class `MyClass` and attribute `__foo`, `__foo` will only be invoked as `_MyClass__foo`
+    * Dunder methods (i.e., double underscores starting and ending a value such as `__name_of_method__`) are magic methods such as `__init__` or `__add__`. They should not be invented, only modified as necessary.
+* Constants should be in ALL CAPS and separated by underscores and defined at the very top of the module if necessary (`EPOCHS = 250`)
+
+
+## Strings
+
+* Whenever possible, use string methods (`'abc'.split()`) as opposed to using the string module (`str.split('String to split')`)
```
Comment (chelseatroy) [inline python.md]: see review

File: python.md [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+if 'pathing'.startswith('path'):
+
+# not good
+if 'pathing'[:3] == 'path':
+```
+     
+## Comments
+
+* Comments should be used sparingly and should ideally answer _why_ a piece of code exists, not _how_ it operates and never _what_ it does
+    * The code itself should tell the reader what it does. If the is too complex to do that, it should be refactored
```
Comment (chelseatroy) [inline python.md]: This is a noble goal, but it hits limitations.

Example: integrating with AWS services via HTTP API. 
The interface has some convoluted requirements to which we must adhere to get the thing onto S3 or what have you, but the integration is what it is, and unfortunately it's not ours to change.
In these cases, I have seen a comment about what the code is doing save other developers a lot of time and headache. 

Another example: we have a frequent breaking point in some of our model notebooks: you start getting index errors halfway down the notebook if your train_test_split is set to 1.0 at the top. We have run into this several times, and every time it's a different developer, so we each spend a chunk of time trying to figure out what the problem is. A comment here explaining a known risk in the way that notebook is structured would be extremely helpful.

File: python.md [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+* Keep comments up-to-date with the code to which they refer. Comments that are misleading are worse than no comments!
+* Comments over multiple lines should be at the same indentation level of the code to which they refer
+* Inline comments should be rare
+
+## Exceptions
+
+* In general, use exceptions sparingly, as they may hide errors that are important. 
+* When creating new exceptions, always derive from `Exception`, not `BaseException`
+* Design exceptions based on what the code that will execute after the exception is raised will need, not based on where the exception occurred. In other words, design exceptions to answer "What went wrong?" versus "A problem occured"
+* When capturing exceptions, use specific exceptions instead of a bare `except:`. The bare `except:` clause will capture `SystemExit` and `KeyboardInterruption` exceptions as well, which can make it harder to kill a running program with `Ctrl-C` and can hide exceptions that we care about.
```
Comment (chelseatroy) [inline python.md]: This is an excellent example of explaining why our convention is what it is!

File: python.md [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+* Comments should be used sparingly and should ideally answer _why_ a piece of code exists, not _how_ it operates and never _what_ it does
+    * The code itself should tell the reader what it does. If the is too complex to do that, it should be refactored
+* Keep comments up-to-date with the code to which they refer. Comments that are misleading are worse than no comments!
+* Comments over multiple lines should be at the same indentation level of the code to which they refer
+* Inline comments should be rare
+
+## Exceptions
+
+* In general, use exceptions sparingly, as they may hide errors that are important. 
+* When creating new exceptions, always derive from `Exception`, not `BaseException`
```
Comment (chelseatroy) [inline python.md]: see review

File: python.md [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+
+* Use `is` or `is not` when comparing a variable to a singleton like `None`
+
+```python
+# good
+if x is None:
+    pass
+    
+# not good
+if x is == None:
```
Comment (chelseatroy) [inline python.md]: Should this be `if x == None`?

File: python.md [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+    pass
+```
+
+* Use `is not` over `if not ... is`
+* Always use `def` instead of `lambda` if you will bind that expression to an identifier. This improves readability of tracebacks. Additionally, assigning a lambda expression to an identifier removes the key benefit of using such an expression (that its definition and use can be embedded in a larger expression).
```
Comment (chelseatroy) [inline python.md]: Also an excellent explanation

File: python.md [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+```python
+# good
+def add_two(val):
+    return val + 2
+    
+# not good
+add_two = lambda val: return val + 2
+```
+
+* Using a context manager (`with`) when acquiring or releasing materials is preferred
```
Comment (chelseatroy) [inline python.md]: see review

File: python.md [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+
+* **Assigning attributes with inherited classes can create unexpected behavior.** When a class inherits from another class, it does not create new attributes but refers back to the inherited classes attributes. This means that changing the attributes within the base class may also change the attributes of the inheriting class as well. To avoid:
+    * For attributes that should differ based on instantiation, define those attributes as `self.attribute` within a `__init__(self):`
+    * For attributes that should remain constant across different instantiations, either define the inherited attribute explicitly (`B.x = 1`) or be aware of how this operates in practice.
+
+```python
+class A:
+    x = 1
+    
+class B:
```
Comment (chelseatroy) [inline python.md]: `class B(A):`

File: python.md [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+    return new_list
+    
+# expected
+>>> append_list(4) # [1, 2, 3, 4]
+
+# actual
+>>> append_list(4) #UnboundLocalError
+```
+
+* In general, removing (or modifying) items of a list while iterating over it is not recommended. It is much better to use list comprehension (`[x for x in range(10) if x % 2 == 0]` versus iterating through the list and removing items) or create a new list and add items to that.
```
Comment (chelseatroy) [inline python.md]: see review

**Comment by richardjonathonharris** [comment]
@chelseatroy thank you so much! I was trying to keep an eye around "good/bad" and "should/should not" but it was tough to catch the bulk of those cases. I'm really excited to incorporate language that is more understanding here.

File: python.md (hunk 1 above)
Comment (richardjonathonharris) [reply python.md]: That is super cool (and certainly something I'm going to bring up at parties from now on!) 

I don't particularly agree with 80 characters either, though some limits are certainly helpful. I could see 100-120 being a good place, with the knowledge that these aren't hard and fast rules but rather helpful(-ish?) guidelines. 

@danilito19 🤔 that's a really good question. I've found working through Parsley that it's been really *really* helpful to have such explicitly named methods and classes (which, while they take up more space, helped me get a sense of the app much faster than I would have otherwise). Maybe we (in the sense of the whole DS&E team) could do some in-person code reviews? I know I could really get some benefit from some group discussion on style and refactoring (though, I also know that putting code up in front of people like that is terrifying in the best of cases).

File: python.md (hunk 2 above)
Comment (richardjonathonharris) [reply python.md]: Agreed. I think we'd run into this less with 100-120 character limit, though!

File: python.md (hunk 3 above)
Comment (richardjonathonharris) [reply python.md]: So that's an interesting point. If I'm in Pandas, I much prefer to see `df['column1']` versus `df["column1"]` (even though both are acceptable). My preference would be to standardize it across the board (always use single or double quotes unless otherwise necessary) but that's overly restrictive.

Another option is that we could get a Python linter as a [git commit hook](https://github.com/pre-commit/pre-commit-hooks) and have it handle some of the formatting concerns for us

File: python.md (hunk 5 above)
Comment (richardjonathonharris) [reply python.md]: 👍

File: python.md (hunk 4 above)
Comment (richardjonathonharris) [reply python.md]: To push back on that a little, even if it's more verbose, I tend to prefer being explicit about what to expect out of a function. To the broader point, I think this is a good case that we should write about to discuss what a good function looks like in Python (i.e., if we expect a boolean to be returned from a function and `under_10` looks like it would do exactly that, we would want to ensure that only a boolean is returned).

File: python.md (hunk 8 above)
Comment (richardjonathonharris) [reply python.md]: 🤔 You bring up a couple of really good cases. I'm gonna have a think on these.

In general though, I agree with the perspective of follow these guidelines until they're a bad idea to follow (which, for the two cases above, probably fit that standard?) 

Maybe one thing to add is to note that these are guidelines, not hard and fast rules. That's always worth repeating.

File: python.md (hunk 14 above)
Comment (richardjonathonharris) [reply python.md]: Python! 🐍

File: python.md [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+# expected
+>>> A.x = 2
+>>> A.x # 2
+>>> B.x # 1
+
+# actual
+>>> A.x = 2
+>>> A.x # 2
+>>> B.x # 2
+```
```
[the remark this answers is not in the document]
Comment (richardjonathonharris) [reply python.md]: I need to rewrite that. It should give the expected weird behavior if you define B as:

```python
class B(A):
    pass
```

File: python.md (hunk 1 above)
Comment (sallee001) [reply python.md]: thumbs up to in-person code reviews and thumbs up to 100-120 characters!

File: python.md (hunk 3 above)
Comment (sallee001) [reply python.md]: I think i generally prefer single quotes for short strings and double quotes for longer strings like sentences or paragraphs, but i don't have a strong preference. the only interaction that comes to mind is if you want to use quotes or an apostrophe within a string then double quotes are easier

File: python.md (hunk 6 above)
Comment (richardjonathonharris) [reply python.md]: 👁‍🗨 (Using this for seen and noted)

File: python.md (hunk 7 above)
Comment (richardjonathonharris) [reply python.md]: 👁‍🗨

File: python.md (hunk 9 above)
Comment (richardjonathonharris) [reply python.md]: 👍

File: python.md (hunk 10 above)
Comment (richardjonathonharris) [reply python.md]: 👁‍🗨

File: python.md (hunk 11 above)
Comment (richardjonathonharris) [reply python.md]: Yes it should be. :)

File: python.md (hunk 12 above)
Comment (richardjonathonharris) [reply python.md]: 👍

File: python.md (hunk 13 above)
Comment (richardjonathonharris) [reply python.md]: 👁‍🗨

File: python.md (hunk 8 above)
Comment (sallee001) [reply python.md]: agree broadly with everything above. in the specific case of notebooks a) i should have written those particular notebooks in a way that avoids those errors which actually supports the point that if you need a comment you probably should think about changing the code, b) we probably need more general style guidelines about notebooks. but yes, sometimes comments are necessary!

**Comment by richardjonathonharris** [comment]
Future plans for this style guide:

* Consider character count
* Review and extra explanation per @chelseatroy 's comments
* Liberally ~pilfer~ cite with attribution the style guide @danilito19 linked [here](https://github.com/kengz/python)
* Review [Khan Academy](https://github.com/Khan/style-guides/blob/master/style/python.md) and [Google](https://github.com/google/styleguide/blob/gh-pages/pyguide.md) style guide
* Develop Numpy, Pandas, Scikit-learn, Keras guide (stretch goal)

File: python.md (hunk 4 above)
Comment (sallee001) [reply python.md]: yeah i think this example is a bit misleading - i would expect such a function to return `False`, not `None`

File: python.md (hunk 8 above)
Comment (richardjonathonharris) [reply python.md]: On point b), guidelines around Jupyter notebook (particularly how to do PRs with them) is certainly something we should fold into this process. 

As for the S3 logic, that seems like the sort of thing that comments would support, particularly if we don't see the integration changing frequently (which would also alleviate the worry of code and comment drifting apart)

File: python.md [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+
+* Comments should be used sparingly and should ideally answer _why_ a piece of code exists, not _how_ it operates and never _what_ it does
+    * The code itself should tell the reader what it does. If the is too complex to do that, it should be refactored
+* Keep comments up-to-date with the code to which they refer. Comments that are misleading are worse than no comments!
+* Comments over multiple lines should be at the same indentation level of the code to which they refer
+* Inline comments should be rare and follow the guidelines above
+
+## Exceptions
+
+* In general, use exceptions sparingly, as they may hide errors that are important
```
Comment (sallee001) [inline python.md]: i'm a little worried about discouraging people from using exceptions - in my experience, data scientists in particular don't use enough exceptions. i think we probably need to have a broader discussion about when to use exceptions and at what level

File: python.md [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+* When creating new exceptions, always derive from `Exception`, not `BaseException`
+* Design exceptions based on what the code that will execute after the exception is raised will need, not based on where the exception occurred. In other words, design exceptions to answer "What went wrong?" versus "A problem occured"
+* When capturing exceptions, use specific exceptions instead of a bare `except:`. The bare `except:` clause will capture `SystemExit` and `KeyboardInterruption` exceptions as well, which can make it harder to kill a running program with `Ctrl-C` and can hide exceptions that we care about. 
+* Limit the amount of code within any given `try` statement so that you avoid hiding bugs
```
Comment (sallee001) [inline python.md]: we should probably mention the "else" and "finally" parts of the try/except control flow - good places to put additional logic after the try block

File: python.md [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+with conn.begin_transaction():
+    do_stuff_in_transaction(conn)
+    
+# not good
+with conn:
+    do_stuff_in_transaction(conn)
+```
+
+* Be consistent in `return` statements
+    * Rule of thumb: either a function should _always_ return something (even `None`) or it should _never_ return anything
```
Comment (sallee001) [inline python.md]: i'm not sure if we want everything here to be python specific or also just generally good coding practices, but this might be a good place to talk about avoiding side effects in functions

File: python.md [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+if thing_that_is_true:
+
+# not good
+if thing_that_is_true == True:
+
+# not good
+if thing_that_is_true is True:
+```
+
+## Typical Python Mistakes
```
Comment (sallee001) [inline python.md]: thanks for adding these in!

File: python.md [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+    * If a method name would clash with an established method, add a trailing underscore (i.e., `class_` instead of `clss`)
+    * If a method is designed to be non-public, start it with an underscore (`_private_method()`). 
+        * This will prevent the method from being imported through `from MyClass import *`
+    * Starting the name with two underscores `__` will invoke name mangling to avoid clashes with any subclasses:
+        * For class `MyClass` and attribute `__foo`, `__foo` will only be invoked as `_MyClass__foo`
+    * Dunder methods (i.e., double underscores starting and ending a value such as `__name_of_method__`) are magic methods such as `__init__` or `__add__`. They should not be invented, only modified as necessary.
+* Constants should be in ALL CAPS and separated by underscores and defined at the very top of the module if necessary (`EPOCHS = 250`)
+
+
+## Strings
```
Comment (sallee001) [inline python.md]: This would be a good place to indicate that we prefer string formatting as `"Some string with a {}".format(value)` rather than the `%` notation

File: python.md [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+    pass
+
+# bad
+def newFunction():
+    pass
+```
+* Classes should be named in Capital Case (`NewClass`)
+* Methods / Attributes should typically follow standards for functions, however:
+    * If a method name would clash with an established method, add a trailing underscore (i.e., `class_` instead of `clss`)
+    * If a method is designed to be non-public, start it with an underscore (`_private_method()`).
```
Comment (sallee001) [inline python.md]: I think it bears emphasizing this point, like "Please ensure all non-public methods start with an underscore" or something along those lines

File: python.md [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+* Multiple statements on the same line are discouraged (avoid `if x: y()`), though, in isolated cases it can be ok. In general, however, it is strongly encouraged to put statements on new lines.
+
+## Imports
+
+* Group imports in the following order, separated by a blank line (alphabetical within each grouping):
+    * Standard Library (`re`, `math`)
+    * Third Party Imports (`pandas`, `scikit-learn`)
+    * Local Application Imports (`parsley.models`, etc.)
+* Absolute imports are highly recommended (`from sklearn.linear_models import LogisticRegression`)
+* However, if required, explicit relative imports are allowed (`from .models import ReferenceDocument`)
```
Comment (sallee001) [inline python.md]: imports in python can be extremely thorny and we may just want to link to some posts about it like this one: https://stanford.edu/~chrisyeh/2017/08/08/definitive-guide-python-imports.html

File: python.md [hunk 24]
```diff
@@ -0,0 +1,309 @@
...
+* Specific libraries sometimes have specific ways to import them:
```
Comment (sallee001) [inline python.md]: I would reword this a bit - it's only by convention that pandas is imported as pd or numpy as np, they don't have to be. We prefer the convention of course

File: python.md [hunk 25]
```diff
@@ -0,0 +1,309 @@
+# Python
+
+Much of this style guide directly follows [PEP 8](http://pep8.org/). For cases that are not covered here, we encourage you to follow those guidelines. Additionally, [PEP 20](https://www.python.org/dev/peps/pep-0020/) contains some helpful pointers as well.
```
Comment (sallee001) [inline python.md]: Can we include other good links either here or at the bottom (for example, hitchhiker's guide to python)?

File: python.md [hunk 26]
```diff
@@ -0,0 +1,309 @@
...
+# good
+def my_new_function():
+    pass
+
+# bad
+def newFunction():
+    pass
+```
+* Classes should be named in Capital Case (`NewClass`)
+* Methods / Attributes should typically follow standards for functions, however:
```
Comment (sallee001) [inline python.md]: At some point we probably want to talk about when to use instance methods vs class methods vs static methods (the answer being - use instance methods unless you have a good reason not to)

File: python.md (hunk 17 above)
Comment (richardjonathonharris) [reply python.md]: I can agree with that. I think we should at least have an example of:

1. Using too few exceptions
2. Using too many exceptions 
3. Using an exception when coding would be more appropriate (such as using an exception to check the type of an object, for example)

File: python.md (hunk 18 above)
Comment (richardjonathonharris) [reply python.md]: Good call. I think that is something that we don't use enough as is (though using it _well_ can be difficult)

File: python.md (hunk 19 above)
Comment (richardjonathonharris) [reply python.md]: I was imaging this style guide to include both, but maybe this is a case where we would want to have a document with good coding practices in general. Could also include things like how to make a good PR at Ascent and keeping git histories clean.

File: python.md (hunk 21 above)
Comment (richardjonathonharris) [reply python.md]: 👍

File: python.md (hunk 22 above)
Comment (richardjonathonharris) [reply python.md]: Agreed. It can be kind of hard to find

File: python.md (hunk 23 above)
Comment (richardjonathonharris) [reply python.md]: Having reviewed a couple of other guides, this is also maybe a section where we want to talk about circular imports as well

File: python.md (hunk 24 above)
Comment (richardjonathonharris) [reply python.md]: 👍

File: python.md (hunk 25 above)
Comment (richardjonathonharris) [reply python.md]: 👍

File: python.md (hunk 26 above)
Comment (richardjonathonharris) [reply python.md]: Absolutely. Also a useful place to talk about `@property`

File: python.md (hunk 24 above)
Comment (chelseatroy) [reply python.md]: On time I saw `import pandas as 🐼 `

I'm not recommending we do that but it was pretty funny

**Comment by danilito19** [comment]
@richardjonathonharris I know there are a ton of comments on this so I wanted to check with you if you were able to implement the ones with general consensus? I want to get this merged soon-ish! No rush tho, obviously not top priority.
